pkg/operator/utils/utils.go

```diff
@@ -292,79 +294,123 @@ func IsValidMode(operatorMode operatorv1.CloudCredentialsMode) bool {
-func UpgradeableCheck(kubeClient client.Client,
-	mode operatorv1.CloudCredentialsMode,
-	toRelease string,
-	newSecrets []types.NamespacedName,
-	rootSecret types.NamespacedName) *configv1.ClusterOperatorStatusCondition {
+func UpgradeableCheck(kubeClient client.Client, mode operatorv1.CloudCredentialsMode, rootSecret types.NamespacedName) *configv1.ClusterOperatorStatusCondition {
 	upgradeableCondition := &configv1.ClusterOperatorStatusCondition{
 		Type: configv1.OperatorUpgradeable,
 	}
 
+	clusterVersion := &configv1.ClusterVersion{}
+	if err := kubeClient.Get(context.TODO(), types.NamespacedName{Name: "version"}, clusterVersion); err != nil {
+		log.Errorf("Failed to get ClusterVersion object while calculating Upgradeable: %s", err)
+		upgradeableCondition.Status = configv1.ConditionFalse
+		upgradeableCondition.Reason = constants.ErrorDeterminingUpgradeableReason
+		upgradeableCondition.Message = fmt.Sprintf("Error getting ClusterVersion while determining upgradability: %s", err)
+		return upgradeableCondition
+	}
+
+	// Note: semver package wants the version string to start with a "v"
+	clusterSemVer := getClusterVersionCompleted(clusterVersion)
+	if clusterSemVer == "" {
+		// cluster has not successfully completed an installation/upgrade at all
+		log.Info("Cluster has not successfully completed an installation/upgrade yet")
+		// don't set any condition for this case
+		return nil
+	}
+
+	if !semver.IsValid(clusterSemVer) {
+		upgradeableCondition.Status = configv1.ConditionFalse
+		upgradeableCondition.Reason = constants.ErrorDeterminingUpgradeableReason
+		upgradeableCondition.Message = fmt.Sprintf("Unable to decode cluster version: %s", clusterSemVer)
+		return upgradeableCondition
+	}
+
+	clusterSemVer = semver.MajorMinor(clusterSemVer)
+
+	// If the cluster admin has set the annotation, then bypass the Upgradeable calculation below and assume
+	// the user knows what they are doing.
+	operatorConfig := &operatorv1.CloudCredential{}
+	operatorConfigKey := types.NamespacedName{
+		Name: constants.CloudCredOperatorConfig,
+	}
+	if err := kubeClient.Get(context.Background(), operatorConfigKey, operatorConfig); err != nil {
+		log.WithError(err).Error("unexpected error checking for CloudCredential config annotation")
+		upgradeableCondition.Status = configv1.ConditionFalse
+		upgradeableCondition.Reason = constants.ErrorDeterminingUpgradeableReason
+		upgradeableCondition.Message = fmt.Sprintf("Error determining if cluster can be upgraded to %s: %s",
+			constants.UpgradeableNextRelease, err)
+		return upgradeableCondition
+	}
+
+	upgradeableTo := operatorConfig.Annotations[constants.UpgradeableAnnotation]
+
+	if !strings.HasPrefix(upgradeableTo, "v") {
+		upgradeableTo = "v" + upgradeableTo
+	}
+
+	upgradeableTo = semver.MajorMinor(upgradeableTo)
+
+	if semver.Compare(upgradeableTo, clusterSemVer) == 1 {
+		log.Info("Found annotation indicating upradeability, assuming upgradeable")
+		return nil
+	}
```

Comment:
Feels like all of this should be guarded on manual mode and moved into the block below.

Reply:
This occurrs before the Manual mode block so that someone can get a cluster unstuck (upgrade-wise) even for non-Manual cases. We can move it into the Manual mode block, but then we lose the escape hatch for all modes that this currently provides.

Reply:
`someone can get a cluster unstuck (upgrade-wise) even for non-Manual cases` I did not understand what this means

Reply:
In the event that our upgradeable code path has an issue (no matter what mode CCO is in), setting the annotation will bypass all the Upgradeable checks and just mark CCO as upgradeable no matter what. This is just a failsafe so that we can get a cluster/customer unstuck in the unlikely ;) event that we mess up.